src/foam/core/CSV.js

```diff
@@ -0,0 +1,359 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.csv',
+  name: 'Outputter',
+
+  documentation: 'CSV Outputter.',
+
+  properties: [
+    {
+      class: 'String',
+      name: 'buf_',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'includeQuotes',
+      help: 'If true, values are quoted with the `"` char',
+      value: true
+    },
+    {
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
+      value: '__'
+    },
+    {
+      class: 'StringArray',
+      name: 'nestedObjectNames'
+    },
+    {
+      class: 'String',
+      name: 'nlStr',
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputHeaderRow',
+      value: true
+    },
+    {
+      class: 'Function',
+      name: 'propertyPredicate',
+      value: function(o, p) { return ! p.transient; }
+    },
+    {
+      class: 'Boolean',
+      name: 'sortObjectKeys',
+      value: false
+    }
+  ],
+
+  methods: [
+    function out() {
+      for ( var i = 0 ; i < arguments.length ; i++ ) { 
+        this.buf_ += arguments[i];
+      }
+
+      return this;
+    },
+
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyFilteredName(o);
+      this.removeTrailingComma().out(this.nlStr);
+    },
+
+    function removeTrailingComma() {
+      // Removes trailing comma
+      if ( this.buf_.charAt(this.buf_.length - 1) == ',' ) this.buf_ = this.buf_.slice(0, -1);
+      return this;
+    },
+
+    function stringify(o) {
+      // Outputs object headers
+      this.outputHeader(o);
+      var ret = this.buf_;
+      this.reset(); // reset to avoid retaining garbage
+
+      // Outputs object values
+      this.output(o);
+      this.removeTrailingComma().out(this.nlStr);
+      ret += this.buf_;
+      this.reset();
+
+      return ret;
+    },
+
+    function externalProperty(o, p) {
+      if ( ! this.propertyPredicate(o, p) ) return false;
+      if ( p.isDefaultValue(o[p.name]) ) return false;
```

Comment:
I think we need to output default values. We don't need to do that for JSON because there we're usually loading the data into another FOAM system which already knows the default values. For CSV though, FOAM is very unlikely to be the recipient. Most like it will be a spreadsheet, which doesn't know the default values.